### PR TITLE
DEP: interpolate: replace interp2d with stub

### LINKED ISSUE #18583: DEP: replace `scipy.interpolate.interp2d` by a stub
Follow-up to #17180 & #17363

The intention here is not (yet) full removal. As suggested on the [mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/thread/LU72CAGYFH37PZWCANMUF6UWFNSPZRHU), the plan is:
- [x] in release 1.10, `interp2d` starts emitting DeprecationWarnings;
- [x] in release 1.14, `interp2d` implementation is replaced by a stub which gives a hard error with an explanation and a pointer to a porting guide;
 - [ ] in release N+X (X ≫ 2), the stub is finally gone, and an attempt to `from scipy.interpolate import interp2d` gives a usual ImportError. This last point may even turn out to be a SciPy 2.0 event.

### PR DESCRIPTION
**Pull request by j-bowhay**
#### Reference issue

closes gh-18583

### CODE REVIEW

File: scipy/interpolate/_interpolate.py [hunk 1]
```diff
@@ -91,8 +89,10 @@ def lagrange(x, w):
 # !! found, get rid of it!
 
 
+# remove in Scipy 2.0.0
```
Comment (j-bowhay) [inline scipy/interpolate/_interpolate.py:92]: Chosen somewhat arbitrarily, feel free to adjust

File: scipy/interpolate/__init__.py [hunk 2]
```diff
@@ -47,7 +47,6 @@
    CloughTocher2DInterpolator
    RBFInterpolator
    Rbf
-   interp2d
```
Comment (j-bowhay) [inline scipy/interpolate/__init__.py:50]: Do we still want this to appear in the documentation?

**Comment by ev-br** [comment]
Did you change your mind Jake on postponing this to 1.13?
https://github.com/scipy/scipy/pull/18022#issuecomment-1439615471

**Comment by j-bowhay** [comment]
Ha you know what I simply forgot. Since it seems we are no closer to being able to update the theme version, which would make the link clickable, I wonder if we just bite the bullet. What do you think?

**Comment by ev-br** [comment]
Either way is fine by me really. This one is going to hurt one way or the other, so can probably pull the plug it sooner than later. If only to gauge the fallout radius.

**Comment by j-bowhay** [comment]
Will reopen for 1.13.0

File: scipy/interpolate/_interpolate.py [hunk 3]
```diff
@@ -113,256 +111,23 @@ def lagrange(x, w):
 
 class interp2d:
     """
-    interp2d(x, y, z, kind='linear', copy=True, bounds_error=False,
-             fill_value=None)
+    `interp2d` was deprecated in SciPy 1.10 and has been removed as of
+    SciPy 1.12.0.
```
Comment (lucascolley) [inline scipy/interpolate/_interpolate.py:115]: Can change this from 12 to 13 now.

Comment (ev-br) [reply scipy/interpolate/_interpolate.py:115]: In fact, I'd consider 1.14 if 1.13 is going to be an out-of-order release to catch up with numpy 2.0. 
What's the plan for 1.13 @tylerjereddy, is it expected around February 2024 or in June 2024?

Comment (tylerjereddy) [reply scipy/interpolate/_interpolate.py:115]: `1.13.0` is expected go out as soon as NumPy `2.0.0` is out, but I'm not sure if the schedule for `2.0.0` is being bumped beyond early February or not.

File: scipy/interpolate/_interpolate.py [hunk 4]
```diff
@@ -96,6 +93,7 @@ def lagrange(x, w):
 # !! found, get rid of it!
 
 
+# remove in Scipy 2.0.0
 dep_mesg = """\
 `interp2d` is deprecated in SciPy 1.10 and will be removed in SciPy 1.13.0.
```
Comment (h-vetinari) [inline scipy/interpolate/_interpolate.py:98]: This message can be removed completely AFAICT (once 1.13 branches)

**Comment by j-bowhay** [comment]
Closing until we branch for 1.14

**Pull request closed without merging**